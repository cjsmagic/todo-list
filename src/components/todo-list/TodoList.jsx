import { useState } from "react";
import TodoItem from "./TodoItem";

function TodoList() {
  const [todo, setTodo] = useState("");
  const [items, setItems] = useState([]);
  const [updateTodo, setUpdateTodo] = useState("");

  return (
    <div className="todo-list" id="124">
      <h1>Todo List</h1>

      <div className="todo-list__input-controls">
        <input
          type="text"
          placeholder="Enter todo"
          value={todo}
          onChange={(e) => setTodo(e.target.value)}
          data-testid="test-id-todo-input"
        />
        <button
          type="button"
          onClick={() => {
            setItems([{ name: todo, id: new Date().getTime() }, ...items]);
            setTodo("");
          }}
        >
          Add
        </button>
      </div>

      <div className="todo-list__items">
        {items.length === 0 && (
          <div className="todo-list__item todo-list__item--no-items">
            <div className="todo-list__item__name">No Items Found</div>
          </div>
        )}

        {items.map((item) => (
          <TodoItem
            key={item.id}
            item={item}
            updateTodo={updateTodo}
            setUpdateTodo={setUpdateTodo}
            items={items}
            setItems={setItems}
          />
        ))}
      </div>
    </div>
  );
}

export default TodoList;
