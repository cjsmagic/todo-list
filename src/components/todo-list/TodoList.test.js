// __tests__/TodoList.test.js
import React from 'react'
import renderer from 'react-test-renderer';
import { render, fireEvent, waitFor, screen, cleanup } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import TodoList from './TodoList'

afterEach(() => {
    cleanup()
})

test('it should load', async () => {
    render(<TodoList />);
    expect(screen.getByRole('heading', { name: /todo list/i })).toHaveTextContent('Todo List')
})

it('renders according to design', () => {
    const tree = renderer
        .create(<TodoList />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});


test('it should add a new item', async () => {
    render(<TodoList />)
    const uniqueName = `${new Date().getTime()}__todo`;

    const inputElement = screen.getByTestId('test-id-todo-input');
    fireEvent.change(inputElement, { target: { value: uniqueName } })

    fireEvent.click(screen.getByRole('button', {
        name: /add/i
    }))

    await waitFor(() => expect(screen.getByText(uniqueName)).toBeTruthy())
    await waitFor(() => expect(inputElement.value).toBe(''))
    // screen.debug();
})


test('it should list items', async () => {
    const { container } = render(<TodoList />)
    const uniqueName = `${new Date().getTime()}__todo`;

    const inputElement = screen.getByTestId('test-id-todo-input');
    fireEvent.change(inputElement, { target: { value: uniqueName } })

    fireEvent.click(screen.getByRole('button', {
        name: /add/i
    }))
    // screen.debug();
    const uniqueName2 = `${new Date().getTime()}__todo`;
    fireEvent.change(inputElement, { target: { value: uniqueName2 } })

    fireEvent.click(screen.getByRole('button', {
        name: /add/i
    }))


    await waitFor(() => {

        const items = Array.from(container.querySelectorAll('.todo-list__item'))
        expect(items.length).toEqual(2)
    })

    // screen.debug();
})



test('it should be deletable', async () => {
    render(<TodoList />)
    const uniqueName = `${new Date().getTime()}__todo`;

    const inputElement = screen.getByTestId('test-id-todo-input');
    fireEvent.change(inputElement, { target: { value: uniqueName } })

    fireEvent.click(screen.getByRole('button', {
        name: /add/i
    }))

    await waitFor(() => expect(screen.getByText(uniqueName)).not.toBeNull());
    // screen.debug();
    fireEvent.click(screen.getByRole('button', {
        name: /delete/i
    }));

    await waitFor(() => expect(screen.queryByText(uniqueName)).toBeNull());

    await waitFor(() => expect(screen.queryByText('No Items Found')).not.toBeNull());
    // screen.debug();
})


test('it should be editable', async () => {
    render(<TodoList />)
    // screen.debug();
    const uniqueName = `${new Date().getTime()}__todo`;

    const inputElement = screen.getByTestId('test-id-todo-input');
    fireEvent.change(inputElement, { target: { value: uniqueName } })

    fireEvent.click(screen.getByRole('button', {
        name: /add/i
    }))

    await waitFor(() => expect(screen.getByText(uniqueName)).not.toBeNull());
    // screen.debug();
    const editButton = screen.getByRole('button', {
        name: /edit/i
    });
    await waitFor(() => expect(editButton).not.toBeNull());
    fireEvent.click(editButton);
    // screen.debug();

    await waitFor(() => expect(editButton).not.toBeNull());
    fireEvent.click(editButton);

    const editElement = screen.getByTestId('test-id-todo-edit');
    const uniqueEditName = `${new Date().getTime()}__todo`;
    fireEvent.change(editElement, { target: { value: uniqueEditName } })
    // screen.debug();
    const updateButton = screen.getByRole('button', {
        name: /update/i
    });
    await waitFor(() => expect(updateButton).not.toBeNull());
    fireEvent.click(updateButton);

    await waitFor(() => expect(screen.getByText(uniqueEditName)).toBeTruthy())
    // screen.debug();
})