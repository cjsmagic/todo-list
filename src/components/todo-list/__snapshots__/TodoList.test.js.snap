// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders according to design 1`] = `
<div
  className="todo-list"
  id="124"
>
  <h1>
    Todo List
  </h1>
  <div
    className="todo-list__input-controls"
  >
    <input
      data-testid="test-id-todo-input"
      onChange={[Function]}
      placeholder="Enter todo"
      type="text"
      value=""
    />
    <button
      onClick={[Function]}
      type="button"
    >
      Add
    </button>
  </div>
  <div
    className="todo-list__items"
  >
    <div
      className="todo-list__item todo-list__item--no-items"
    >
      <div
        className="todo-list__item__name"
      >
        No Items Found
      </div>
    </div>
  </div>
</div>
`;
